#### 선형판별분석   
이미 분류되어 있는 집단 사이의 차이를 설명할 수 있는 독립변수를 찾아내고, 이 변수들의 선형결합으로 이루어진 판별식 확인  
H0(귀무가설):방문빈도와 Monetary, 거래기간은 이탈에 영향을 주지않는다.   
H1(대립가설):방문빈도와 Monetary, 거래기간은 이탈에 영향을 준다.   
전제조건  
1. 독립변수는 정규분포를 따라야 한다.  
2. 종속변수와 독립변수는 상관관계를 가져야 한다.   
3. 독립변수들 사이에 상관관계가 없거나 작아야한다.   
4. 종속변수로 구분되는 각 집단별 공분산 행렬이 유사해야한다.

In [3]:
import pandas as pd
import numpy as np
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
from sklearn.metrics import confusion_matrix

In [5]:
df = pd.read_csv('datasets/Ashopping.csv', sep=',', encoding='CP949')

In [6]:
X=df[['방문빈도','Monetary','거래기간']]
Y=df[['이탈여부']]

In [8]:
# 선형판별분석
lda = LDA().fit(X, Y.values.ravel())
lda

LinearDiscriminantAnalysis()

In [17]:
#예측결과
print('판별식 선형계수:',np.round(lda.coef_,3))
print('\n판별식 절편:',np.round(lda.intercept_,3))
print('\n예측결과:\n', pd.DataFrame(lda.predict(X)))
print('\n예측스코어:\n',pd.DataFrame(lda.predict_proba(X)))
print('\n예측정확도:',lda.score(X,Y))

판별식 선형계수: [[ 0.074  0.222 -0.097]]

판별식 절편: [72.999]

예측결과:
      0
0    0
1    1
2    0
3    0
4    0
..  ..
995  0
996  0
997  0
998  0
999  0

[1000 rows x 1 columns]

예측스코어:
                 0             1
0    1.000000e+00  2.471028e-13
1    1.185558e-10  1.000000e+00
2    1.000000e+00  1.549378e-13
3    1.000000e+00  1.011901e-10
4    1.000000e+00  1.542091e-11
..            ...           ...
995  1.000000e+00  1.895547e-12
996  1.000000e+00  2.153449e-10
997  1.000000e+00  2.691201e-14
998  1.000000e+00  8.726427e-13
999  1.000000e+00  6.063388e-13

[1000 rows x 2 columns]

예측정확도: 0.986


In [19]:
#분류행렬표 출력
cf_m = pd.DataFrame(confusion_matrix(Y, lda.predict(X)))
cf_m.colums=['예측0', '예측1']
cf_m.index=['실제0', '실제1']
cf_m

C:\Users\Public\Documents\ESTsoft\CreatorTemp/ipykernel_43440/2061594439.py:3: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  cf_m.colums=['예측0', '예측1']


0   1
실제0  921  12
실제1    2  65

예측 정확도는 0.986으로 매우 높은 정확도를 보였다.  
선형판별분석 결과 도출된 선형계수와 절편을 통해 선형판별식은  
Y=72.999 + 0.074*방문빈도 + 0.222*Monetary - 0.097*거래기간 으로 알수있다.  